# Amazon Reviews 

## Task 1 :

We have three text files containg reviews. From these datasets, find the one that in most suitable for sentiment analysis based on the following factor.

1. The data bias should be minimal.


In [1]:
file1 = open("D:\Data Science\Course 12\ds3_c2_s3_hackathon-master\ds3_c2_s3_hackathon-master\Review1.txt",'r').read()

In [2]:
file2 = open("D:\Data Science\Course 12\ds3_c2_s3_hackathon-master\ds3_c2_s3_hackathon-master\Review2.txt",'r').read()

In [3]:
file3 = open("D:\Data Science\Course 12\ds3_c2_s3_hackathon-master\ds3_c2_s3_hackathon-master\Review3.txt",'r').read()

In [4]:
import spacy
from spacy.lang.en import English
nlp = spacy.load('en_core_web_sm')
import collections
from typing import Dict,List, Tuple

In [5]:
import collections
from typing import Dict,List, Tuple

In [6]:
doc1 = nlp(file1)

In [13]:
def count_sentiment_words(text):
    docs=nlp(text)
    text=" ".join(str(doc) for doc in docs if(doc.is_stop==False))
    good=['good', 'great', 'excellent', 'positive', 'wonderful', 'amazing', 'fantastic', 'terrific', 'superb', 'outstanding']
    bad=['bad', 'terrible', 'awful', 'horrible', 'negative', 'sad', 'depressing', 'unhappy', 'gloomy', 'bleak']
    neutral=['neutral', 'okay', 'average', 'ordinary', 'typical', 'commonplace', 'mundane', 'routine', 'standard', 'usual']
    good_count=sum([text.lower().count(word) for word in good])
    bad_count=sum([text.lower().count(word) for word in bad])
    neutral_count=sum([text.lower().count(word) for word in neutral])

    return {'good': good_count, 'bad': bad_count, 'neutral': neutral_count}
sentiments=count_sentiment_words(doc1)
print('The proportion of words in each sentiment : ')
print(sentiments)

The proportion of words in each sentiment : 
{'good': 859, 'bad': 160, 'neutral': 69}


In [14]:
doc2 = nlp(file2)

In [15]:
sentiments=count_sentiment_words(doc2)
print('The proportion of words in each sentiment : ')
print(sentiments)

The proportion of words in each sentiment : 
{'good': 733, 'bad': 180, 'neutral': 60}


In [16]:
doc3 = nlp(file3)

In [17]:
sentiments=count_sentiment_words(doc3)
print('The proportion of words in each sentiment : ')
print(sentiments)

The proportion of words in each sentiment : 
{'good': 723, 'bad': 155, 'neutral': 56}


## Interpretation :  

- file 1 : {'good': 859, 'bad': 160, 'neutral': 69}
- file 2 : {'good': 733, 'bad': 180, 'neutral': 60}
- file 3 : {'good': 723, 'bad': 155, 'neutral': 56}

File 2 is suitable as file2 have less data bias  


## Task  2 :

Follow the steps to clean and preprocess the text file that you find most suitable.

1. Remove stop words. 
2. Remove the following words : 'ingredients', 'live', 'person', 'went','came', 'gave', 'seen' and 'movie'. Verify that the mentioned stop words have been removed from the input document.
3. Remove all the unnecessary symbols such as $, ... , !!! , ---, etc
4. Remove '__label__1', '__label__2', 
5. Apply lemmatization.
6. Apply part-of-speech(POS) tagging.
7. Make the text ready for analysis in a format That the machine can understand

In [35]:
import re
labels=['__label__1', '__label__2', 'label__1', 'label__2']
tokens=[token for token in nlp(file2) if not (token.is_stop or str(token) in labels)]


In [36]:
tokens

[_,
 _,
 Goes,
 ",
 Haywire,
 ",
 Regularly,
 :,
 Setup,
 easy,
 ,,
 printer,
 routinely,
 goes,
 haywire,
 ,,
 spitting,
 cartridge,
 load,
 paper,
 refusing,
 print,
 .,
 tried,
 different,
 PCs,
 ,,
 intermittent,
 weird,
 behavior,
 .,
 reset,
 printerand,
 computer,
 fix,
 .,
 I'm,
 buy,
 Epson,
 !,
 ,
 _,
 _,
 Old,
 fashioned,
 ,,
 sweet,
 story,
 .,
 :,
 challenge,
 read,
 .,
 Enjoyed,
 .,
 change,
 modern,
 world,
 ,,
 recommend,
 .,
 ,
 _,
 _,
 Stay,
 Away,
 :,
 plain,
 bad,
 .,
 Boring,
 .....,
 find,
 bit,
 entertaining,
 interesting,
 .,
 waste,
 time,
 .,
 ,
 _,
 _,
 worst,
 .........,
 movie,
 ............,
 :,
 break,
 open,
 red,
 light,
 camera,
 steal,
 film,
 .,
 Watch,
 .,
 10,000,
 times,
 better,
 acting,
 coherent,
 plot,
 !,
 ,
 _,
 _,
 Blu,
 -,
 Ray,
 Review,
 :,
 terrible,
 movie,
 ,,
 garbage,
 .,
 totally,
 sucks,
 .,
 Hollywood,
 coming,
 .,
 running,
 stories,
 movies,
 .,
 garbage,
 .,
 Good,
 Lord,
 .,
 Right,
 Boondock,
 Saints,
 2,
 Smokin,
 Aces,
 2,


In [40]:
text = ' '.join(str(token) for token in tokens)

In [41]:
import re
clean_text = re.sub(r'[^a-zA-Z]+', ' ',text)

In [42]:
clean_text

' Goes Haywire Regularly Setup easy printer routinely goes haywire spitting cartridge load paper refusing print tried different PCs intermittent weird behavior reset printerand computer fix I m buy Epson Old fashioned sweet story challenge read Enjoyed change modern world recommend Stay Away plain bad Boring find bit entertaining interesting waste time worst movie break open red light camera steal film Watch times better acting coherent plot Blu Ray Review terrible movie garbage totally sucks Hollywood coming running stories movies garbage Good Lord Right Boondock Saints Smokin Aces Nicolas Cage film redeemable quality film Period Nicolas Cage barrel absolutely range acting worse low budget B movie think seriously New Orleans Post Katrina Nic supposed tough guy gets kinds trouble figure coming going baffled entire movie lines HORRIBLE SAVE MONEY watch dirty cop horrible person Spoiler worked end heck point movie Seriously worked end I know continued watch Great Writing think review nee

In [43]:
doc = nlp(clean_text)

In [44]:
print(doc)

 Goes Haywire Regularly Setup easy printer routinely goes haywire spitting cartridge load paper refusing print tried different PCs intermittent weird behavior reset printerand computer fix I m buy Epson Old fashioned sweet story challenge read Enjoyed change modern world recommend Stay Away plain bad Boring find bit entertaining interesting waste time worst movie break open red light camera steal film Watch times better acting coherent plot Blu Ray Review terrible movie garbage totally sucks Hollywood coming running stories movies garbage Good Lord Right Boondock Saints Smokin Aces Nicolas Cage film redeemable quality film Period Nicolas Cage barrel absolutely range acting worse low budget B movie think seriously New Orleans Post Katrina Nic supposed tough guy gets kinds trouble figure coming going baffled entire movie lines HORRIBLE SAVE MONEY watch dirty cop horrible person Spoiler worked end heck point movie Seriously worked end I know continued watch Great Writing think review need

### Remove stop words 

In [45]:
import spacy
en = spacy.load('en_core_web_sm')
stop_words = en.Defaults.stop_words

In [46]:

filter_sent = [ ]

In [47]:
en.Defaults.stop_words |= {"Note", "poster", "ingredients", "live", "person", "went","came", "gave", "seen","movie","label"}

In [48]:
for word in doc:
    if word.is_stop == False:
        filter_sent.append(word)

In [49]:
filter_sent

[ ,
 Goes,
 Haywire,
 Regularly,
 Setup,
 easy,
 printer,
 routinely,
 goes,
 haywire,
 spitting,
 cartridge,
 load,
 paper,
 refusing,
 print,
 tried,
 different,
 PCs,
 intermittent,
 weird,
 behavior,
 reset,
 printerand,
 computer,
 fix,
 m,
 buy,
 Epson,
 Old,
 fashioned,
 sweet,
 story,
 challenge,
 read,
 Enjoyed,
 change,
 modern,
 world,
 recommend,
 Stay,
 Away,
 plain,
 bad,
 Boring,
 find,
 bit,
 entertaining,
 interesting,
 waste,
 time,
 worst,
 movie,
 break,
 open,
 red,
 light,
 camera,
 steal,
 film,
 Watch,
 times,
 better,
 acting,
 coherent,
 plot,
 Blu,
 Ray,
 Review,
 terrible,
 movie,
 garbage,
 totally,
 sucks,
 Hollywood,
 coming,
 running,
 stories,
 movies,
 garbage,
 Good,
 Lord,
 Right,
 Boondock,
 Saints,
 Smokin,
 Aces,
 Nicolas,
 Cage,
 film,
 redeemable,
 quality,
 film,
 Period,
 Nicolas,
 Cage,
 barrel,
 absolutely,
 range,
 acting,
 worse,
 low,
 budget,
 B,
 movie,
 think,
 seriously,
 New,
 Orleans,
 Post,
 Katrina,
 Nic,
 supposed,
 tough,
 guy,


### Lemmatization

In [50]:
for word in filter_sent :
    print(word.text, '-->',word.lemma_)

  -->  
Goes --> go
Haywire --> Haywire
Regularly --> Regularly
Setup --> Setup
easy --> easy
printer --> printer
routinely --> routinely
goes --> go
haywire --> haywire
spitting --> spit
cartridge --> cartridge
load --> load
paper --> paper
refusing --> refuse
print --> print
tried --> try
different --> different
PCs --> pc
intermittent --> intermittent
weird --> weird
behavior --> behavior
reset --> reset
printerand --> printerand
computer --> computer
fix --> fix
m --> m
buy --> buy
Epson --> Epson
Old --> Old
fashioned --> fashioned
sweet --> sweet
story --> story
challenge --> challenge
read --> read
Enjoyed --> Enjoyed
change --> change
modern --> modern
world --> world
recommend --> recommend
Stay --> stay
Away --> away
plain --> plain
bad --> bad
Boring --> Boring
find --> find
bit --> bit
entertaining --> entertain
interesting --> interesting
waste --> waste
time --> time
worst --> bad
movie --> movie
break --> break
open --> open
red --> red
light --> light
camera --> camera


### part-of-speech(POS) tagging

In [51]:
for word in filter_sent:
    print(word.text, '\t', word.pos_, '\t', word.tag_)

  	 SPACE 	 _SP
Goes 	 VERB 	 VBZ
Haywire 	 PROPN 	 NNP
Regularly 	 PROPN 	 NNP
Setup 	 PROPN 	 NNP
easy 	 ADJ 	 JJ
printer 	 NOUN 	 NN
routinely 	 ADV 	 RB
goes 	 VERB 	 VBZ
haywire 	 NOUN 	 NN
spitting 	 VERB 	 VBG
cartridge 	 NOUN 	 NN
load 	 NOUN 	 NN
paper 	 NOUN 	 NN
refusing 	 VERB 	 VBG
print 	 NOUN 	 NN
tried 	 VERB 	 VBD
different 	 ADJ 	 JJ
PCs 	 NOUN 	 NNS
intermittent 	 ADJ 	 JJ
weird 	 ADJ 	 JJ
behavior 	 NOUN 	 NN
reset 	 NOUN 	 NN
printerand 	 PROPN 	 NNP
computer 	 NOUN 	 NN
fix 	 NOUN 	 NN
m 	 AUX 	 VBP
buy 	 VERB 	 VBP
Epson 	 PROPN 	 NNP
Old 	 PROPN 	 NNP
fashioned 	 ADJ 	 JJ
sweet 	 ADJ 	 JJ
story 	 NOUN 	 NN
challenge 	 NOUN 	 NN
read 	 AUX 	 VBD
Enjoyed 	 PROPN 	 NNP
change 	 VERB 	 VBP
modern 	 ADJ 	 JJ
world 	 NOUN 	 NN
recommend 	 VERB 	 VBP
Stay 	 VERB 	 VB
Away 	 ADV 	 RB
plain 	 ADV 	 RB
bad 	 ADJ 	 JJ
Boring 	 PROPN 	 NNP
find 	 VERB 	 VBP
bit 	 NOUN 	 NN
entertaining 	 VERB 	 VBG
interesting 	 ADJ 	 JJ
waste 	 NOUN 	 NN
time 	 NOUN 	 NN
worst 	 ADJ 	 JJS


In [52]:
text = ' '.join(str(token) for token in filter_sent)

In [54]:
text

'  Goes Haywire Regularly Setup easy printer routinely goes haywire spitting cartridge load paper refusing print tried different PCs intermittent weird behavior reset printerand computer fix m buy Epson Old fashioned sweet story challenge read Enjoyed change modern world recommend Stay Away plain bad Boring find bit entertaining interesting waste time worst movie break open red light camera steal film Watch times better acting coherent plot Blu Ray Review terrible movie garbage totally sucks Hollywood coming running stories movies garbage Good Lord Right Boondock Saints Smokin Aces Nicolas Cage film redeemable quality film Period Nicolas Cage barrel absolutely range acting worse low budget B movie think seriously New Orleans Post Katrina Nic supposed tough guy gets kinds trouble figure coming going baffled entire movie lines HORRIBLE SAVE MONEY watch dirty cop horrible person Spoiler worked end heck point movie Seriously worked end know continued watch Great Writing think review needs 